lxd/daemon: Manage images and backups volumes without symlinks

skozina opened the pull request:
When `storage.backups_volume` or `storage.images_volume` local configs are set, don't create any symlinks for `LXD_DIR/backups` or `LXD_DIR/images`. Instead, adjust LXD so that the backup or image files are directly operated at the target volume.

tomponline reviewed (changes requested):
Thanks @skozina 

I did a quick first pass of this but did not go into depth yet.

My initial feedback is:

1. Upgrades/downgrades of LXD: Does this support upgrading from an earlier version of LXD to this version seamlessly? If not then we will need a patch, take a look at `lxd/patches.go`. Specifically the calls to `patchGenericStorage` which allow patches to be called in the storage subsystem.
2. Related to 1. for downgrades, we need to think about whether we want to support downgrading LXD back to an earlier version and for this to work? If not then we need to ensure that this patch-set (or some of it) is not backported to the LTS series (where downgrades are allowed). I would ask that the commits that shouldnt be backported are labelled that way in the commit message somehow for me to see this when backporting.
3. Related to 2. for these reason of potentially not wanting to backport all of this, and to make it easier for this PR to be reviewed I would ask that the changes are broken into many more smaller commits, ideally one for each logical change. For example your work on improving the path function helpers could perhaps be backported, as well as other stylistic or comment fixes, but if we decide not to backport the bulk of the PR then we would not want those changes made within the same commit as the code that changes the on-disk files.

skozina commented:
Thanks for your review, I'll work on addressing your other comments (notably breaking up this work into multiple commits) separately.

> 1. Upgrades/downgrades of LXD: Does this support upgrading from an earlier version of LXD to this version seamlessly? If not then we will need a patch, take a look at `lxd/patches.go`. Specifically the calls to `patchGenericStorage` which allow patches to be called in the storage subsystem.

We are not changing the layout of actual files on the storage (I'm leaving the move of backups/instances into a sub-folder for separate PR). In other words, the full path to image/backup after resolving symlinks stays the same. So there's no need to move files around for this PR.
What is changing is that the "new" LXD version doesn't need the symlink. So as a patch, we could remove this symlink. But leaving it around does not cause any harm. Do you want me to remove the symlink? (read further first)

> 2. Related to 1. for downgrades, we need to think about whether we want to support downgrading LXD back to an earlier version and for this to work? If not then we need to ensure that this patch-set (or some of it) is not backported to the LTS series (where downgrades are allowed). I would ask that the commits that shouldnt be backported are labelled that way in the commit message somehow for me to see this when backporting.

This work is in preparation for the per-project storage. That, as a new feature, doesn't need to land in LTS series I think. On the other hand I'd love to support downgrades anyway, just in case of any issues on update.
If we leave the symlink around on update (see previous comment), the simple "update & downgrade" would work. What would not work is "upgrade, set the config.images_volume setting and downgrade", as that would remove the symlink on the config setting.
To fix this other case, we'd need to detect downgrade and re-create the symlinks as needed. Do we have some pre-downgrade hooks to use?

> 3. Related to 2. for these reason of potentially not wanting to backport all of this, and to make it easier for this PR to be reviewed I would ask that the changes are broken into many more smaller commits, ideally one for each logical change. For example your work on improving the path function helpers could perhaps be backported, as well as other stylistic or comment fixes, but if we decide not to backport the bulk of the PR then we would not want those changes made within the same commit as the code that changes the on-disk files.

Agreed, let me separate the new functions, their calls and then the actual change of the storage path into separate commits. Thanks for the review!

skozina commented on lxd/daemon.go:
```diff
@@ -723,6 +724,24 @@ func (d *Daemon) State() *state.State {
+		return storageDrivers.GetVolumeMountPath(poolName, storageDrivers.VolumeTypeCustom, volStorageName)
+	}
+
+	s.ImagesStoragePath = func() string {
```
Technically this could be done as these functions just operate strings. It's not that simple though, as we can't just add `import storageDrivers` into the `state.State` class, as that would create circular dependency. We'd need this import to get symbols like `GetVolumeMountPath()` and `VolumeTypeCustom`.
We could refactor all these functions into a new package, something like `github.com/lxd/storage/utils`, but that would be a major change, so just wanted to check if that's really worth it. I don't see much other options (other than copy these functions into the `state` package, which would be even worse). WDYT @tomponline  ?

skozina replied:
Lets leave this as it is for now due to the reasons stated above.

skozina commented on lxd/storage/backend_lxd.go:
```diff
@@ -5996,6 +5996,12 @@ func (b *lxdBackend) RenameCustomVolume(projectName string, volName string, newV
 	defer l.Debug("RenameCustomVolume finished")
 
+	// Silence the static analysis tool
+	newVolName, err := backup.ValidateBackupName(newVolName)
```
Thanks, changed to `validVolumeName` as this is local variable.

tomponline replied:
Ah you misunderstood me, i was referring to your use of ValidateBackupName() function when you were validating a custom volume name.

tomponline replied:
We have a function called ValidVolumeName

skozina replied:
Ah! Thanks, makes perfect sense now. Fixed as you suggest.

tomponline replied:
Thanks. I missed the use of the function name on the 2nd pass, so Im glad you queried about the variable name nit :)

skozina commented on lxd/api_1.0.go:
```diff
@@ -10,7 +10,7 @@ import (
 	"slices"
 
-	"github.com/canonical/lxd/client"
+	lxd "github.com/canonical/lxd/client"
```
Thanks for the tips, alias removed from all files.

skozina commented on lxd/api_1.0.go:
I've isolated the variable renames, introduction of new functions, their actual use, and the symlinks changes into separate commits.

skozina commented on lxd/daemon_images.go:
```diff
@@ -409,7 +409,7 @@ func ImageDownload(r *http.Request, s *state.State, op *operations.Operation, ar
 			Canceler:        canceler,
 			DeltaSourceRetriever: func(fingerprint string, file string) string {
-				path := shared.VarPath("images", fmt.Sprintf("%s.%s", fingerprint, file))
+				path := fmt.Sprintf("%s/%s.%s", destDir, fingerprint, file)
```
Thanks, removed all calls to `Sprintf()`, and also isolated this in a separate commit.

simondeziel commented on lxd/images.go:
```diff
@@ -4467,8 +4477,9 @@ func imageExportPost(d *Daemon, r *http.Request) response.Response {
 		createArgs := &lxd.ImageCreateArgs{}
-		imageMetaPath := shared.VarPath("images", details.imageFingerprintPrefix)
-		imageRootfsPath := shared.VarPath("images", details.imageFingerprintPrefix+".rootfs")
+		imageMetaPath := s.ImagesStoragePath()
+		imageMetaPath = filepath.Join(imageMetaPath, details.imageFingerprintPrefix)
```
I don't know if the compiler is smart enough to optimize away the first assignment and do just:

```suggestion
		imageMetaPath := filepath.Join(s.ImagesStoragePath(), details.imageFingerprintPrefix)
```
but I find this more compact/straight forward version easier to read. Probably a question of style/personal taste however :)

skozina replied:
Thanks for the review! I've changed the code as you suggest wherever there was only one reference to the new variable. If it's used on multiple places, I've left the code as it was.

simondeziel replied:
Thanks, you've struck the perfect balance IMHO ;)

simondeziel commented on test/suites/serverconfig.sh:
```diff
@@ -64,8 +64,8 @@ _server_config_storage() {
-  BACKUPS_AFTER=$(find "${LXD_DIR}/backups/" | sort)
-  IMAGES_AFTER=$(find "${LXD_DIR}/images/" | sort)
+  BACKUPS_AFTER=$(cd "${LXD_DIR}/storage-pools/${pool}/custom/default_backups/" && find . | sort)
+  IMAGES_AFTER=$(cd "${LXD_DIR}/storage-pools/${pool}/custom/default_images/" && find . | sort)
```
I wonder if it would make sense to compare the device numbers backing those directories BEFORE and AFTER  changing their `storage.backups_volume`/`storage.images_volume` config? This way, we'd be certain their content was moved. For the dev numbers, I was thinking of something like `stat -c "%d%D"`.

This check should be skipped for the `dir` driver as the backing device would be the same.

Anyway, just a thought :)

skozina replied:
Lets leave this as separate improvement. I've filled this: https://github.com/canonical/lxd/issues/15524 . Thanks!

tomponline commented on lxd/storage/backend_lxd.go:
```diff
@@ -5996,11 +5996,17 @@ func (b *lxdBackend) RenameCustomVolume(projectName string, volName string, newV
 	defer l.Debug("RenameCustomVolume finished")
 
+	// Silence the static analysis tool
+	validVolumeName, err := backup.ValidateBackupName(newVolName)
```
Just a nit, but you could have done:

```go
newVolName, err := backup.ValidateBackupName(newVolName)
```

To keep the variable name `newVolName` which is more accurate than `validVolumeName` (which is it old or new?) and hopefully the static analysis tools would still be happy that its been validated.

skozina replied:
I'm confused, I've changed the variable name based on your previous feedback. I've unresolved the previous conversation, which variable name would you prefer?

tomponline replied:
Just a misunderstanding https://github.com/canonical/lxd/pull/15482#discussion_r2078318810

tomponline commented on lxd/sys/fs.go:
```diff
@@ -73,26 +73,55 @@ func (s *OS) initDirs() error {
+		for _, dir := range dirs {
+			err := os.Mkdir(dir.path, dir.mode)
+			if err != nil {
+				if !os.IsExist(err) {
```
> New code should use errors.Is(err, fs.ErrExist)

https://pkg.go.dev/os#IsExist

tomponline reviewed (changes requested):
A few minor comments.

Also for the future PR that covers on-disk storage layout changes, I found `patchMoveBackupsInstances` which you can probably use for inspiration.

tomponline commented:
@skozina also this would be a good time to review the tests we have for this functionality and check its checking for setting/unsetting and moving existing files back and forth.

skozina commented on lxd/sys/fs.go:
```diff
@@ -73,26 +73,55 @@ func (s *OS) initDirs() error {
 
-	for _, dir := range dirs {
-		err := os.Mkdir(dir.path, dir.mode)
+	if config.StorageBackupsVolume() == "" {
```
Importing the `state` package would cause an import cycle between `sys` and `state` packages. I'm not sure what's the best way to resolve this cycle, can you advice?

tomponline replied:
Ah OK. How about we move the 2 functions into dedicated functions in the `lxd/storage/drivers/utils.go` file?

The `lxd/storage/drivers` package is already imported in a bunch of places as:

```
	storageDrivers "github.com/canonical/lxd/lxd/storage/drivers"
```

We could potentially put them in "github.com/canonical/lxd/lxd/storage/utils.go" but then I worry we will need these paths in the storage drivers later (and that would create a circular dependency again). So this is why I'm thinking storage drivers.

One thing to keep in mind though is that the storage drivers package doesn't know about the database (it shouldnt need to), so in the future when we want to perform project level logic, we'll want to update those functions to accept not only node-local config, but also the project level config arguments.

skozina replied:
If we move the two functions to `lxd/storage/drivers` package, there will again be an import cycle between the `lxd/storage/drivers` and `lxd/backup` packages. In `lxd/storage/drivers` package, implementations of `CreateVolumeFromBackup()` depend on `backup.Info`. And in the `lxd/backup` we'd need to add the dependency on `lxd/storage/drivers` to give the backup functions the new storage path.

skozina replied:
The only option I can see is to create a new package, perhaps `lxd/storage/utils`. It will be a bit confusing as we'll have `lxd/storage/utils.go` (package storage), `lxd/storage/drivers/utils.go` (package drivers), and `storage/utils/utils.go` (package utils). What do you think?

tomponline replied:
Hrm tricky. Try github.com/canonical/lxd/lxd/daemon seems like an appropriate place for daemon level helpers.

tomponline replied:
I tend to try and avoid/reduce our util packages 

https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common

skozina replied:
Hm... no matter where we'd put these common functions (`ImagesStoragePath()` and `BackupStoragePath()`), these need to get the mount point of the volume which is used as backup/images store. We currently use `GetVolumeMountPath()` function from `lxd/lxd/storage/drivers` package for this, here:
https://github.com/canonical/lxd/pull/15482/commits/e91d5839702fb2bef7661a777d87413cb8d95237#diff-eed70a8014a7d0c46b4e437230b15c1afe8f90752f79e48ba983128b3457c1dbR734
This again creates an import cycle, the `daemon` package needs to import `storageDrivers`, but the `storageDrivers` need `archive` and `apparmor`, which again need `daemon`. This cycle:
```
package github.com/canonical/lxd/lxd
        imports github.com/canonical/lxd/lxd/apparmor from daemon.go    
        imports github.com/canonical/lxd/lxd/daemon from archive.go      # New dependency in this PR to update apparmor profiles                                                                                                                                                                   
        imports github.com/canonical/lxd/lxd/storage/drivers from daemon.go    # to get GetVolumeMountPath()                                                                                                                                                           
        imports github.com/canonical/lxd/lxd/archive from driver_btrfs_volumes.go                                                                                                                                                     
        imports github.com/canonical/lxd/lxd/apparmor from archive.go: import cycle not allowed   
```
The `GetVolumeMountPath()` feels like it should stay with `storageDrivers`.
Considering this, I think the current PR which duplicates the function at the one place at `lxd/sys/fs.go` is the best option.

tomponline replied:
I see, its ashame but agree lets go with that for now as this is dragging.

tomponline replied:
Maybe it'll make sense to move more of the image and backup management stuff into the storage package in later iterations and then this will solve itself.

skozina replied:
Thanks. It's frustrating these are all just simple string manipulation function. But I don't see better option for now. That said, I think that all comments are somewhat resolved in this PR.

skozina commented:
@tomponline thanks for your review! After some thinking I've added a patch to remove symlinks on LXD update. There's no point in keeping these around forever. The config setting tests in `suites/serverconfig.sh` seem fairly thorough, I've added a brief test for unsetting of the config values.
See above my comments on your other comments. Thank you!

skozina commented on test/suites/serverconfig.sh:
```diff
@@ -97,10 +97,31 @@ _server_config_storage() {
+    false
+  fi
+
+  lxc start foo
+  lxc exec foo -- touch /root/foo
+  lxc stop -f foo
```
Err.... No, this doesn't seem to make any sense. Let me remove these lines. Thanks!

tomponline commented:
Will take this after the LXD 6.4 tag is created.

skozina commented:
> @skozina this is ready to be rebased now that #15674 is merged.
> 
> You will likely need to update the tests to account for the new location of the image files.

Thanks for the ping, I've rebased this PR.
I'm not sure I follow the test update -- this PR updates the tests as needed: https://github.com/canonical/lxd/pull/15482/commits/58b017b58bb4fea3d396f15382bca830524a470e

I think that #15674 doesn't change any image file location?

tomponline commented:
> Thanks for the ping, I've rebased this PR. I'm not sure I follow the test update -- this PR updates the tests as needed: [58b017b](https://github.com/canonical/lxd/commit/58b017b58bb4fea3d396f15382bca830524a470e)
> 
> I think that #15674 doesn't change any image file location?

You're correct my PR doesn't change the image file location, but it does add tests that check the image files are stored in `LXD_DIR/images/` which I think after your change will potentially break some tests as the symlink wont be being populated anymore.

That being said, it maybe that the cluster image refresh test in question doesn't operate with the nodes operating with a custom storage volume for images. In which case I would suggest we update the test to ensure that one node has image storage set to a custom image volume and then update the test to check for the respective location.

skozina commented:
Added tests in the clustering test suite with one node setup with dedicated image storage. Eventually I've decided to add this to the `clustering_image_replication()` test rather than the `clustering_image_refresh()` test. Thanks!

tomponline commented on lxd/sys/fs.go:
```diff
@@ -73,26 +75,55 @@ func (s *OS) initDirs() error {
+				}
+
+				err = os.Chmod(dir.path, dir.mode)
+				if err != nil && !os.IsNotExist(err) {
```
IsNotExist is deprecated

https://pkg.go.dev/os#IsNotExist

> This function predates [errors.Is](https://pkg.go.dev/errors#Is). It only supports errors returned by the os package. New code should use errors.Is(err, fs.ErrNotExist).

tomponline commented on lxd/patches.go:
```diff
@@ -1487,4 +1488,33 @@ func patchOIDCGroupsClaimScope(_ string, d *Daemon) error {
+	}
+
+	for _, dir := range dirs {
+		if !shared.PathExists(dir) {
```
do we need this extra syscall, can we instead detect the not exist error from `os.Lstat()` instead?

tomponline commented on test/suites/serverconfig.sh:
```diff
@@ -63,6 +63,10 @@ _server_config_storage() {
   lxc config set storage.images_volume "${pool}/images"
 
+  # Validate the old location is really gone
+  [ ! -e "${LXD_DIR}/backups/" ]
```
should we lose the trailing slashes on these? Not sure if this is done intentionally?

tomponline commented:
Thanks! LGTM

simondeziel commented:
I just noticed that one of the (side?) effect of this PR was to close https://github.com/canonical/lxd/security/code-scanning/47

skozina commented:
> I just noticed that one of the (side?) effect of this PR was to close [Uncontrolled data used in path expression](/canonical/lxd/security/code-scanning/47)

Interesting. I have no idea how.
